### PR TITLE
Extend content negotiation functionality

### PR DESCRIPTION
**Pull request by joe94**
This PR adds convenience functions for a WRP handler's content negotiation strategy.

### CODE REVIEW

**Review by kcajmagic — APPROVED** [review]

File: wrphttp/requestResponse.go [hunk 1]
```diff
@@ -71,7 +92,7 @@ type ResponseWriter interface {
 	// WriteWRP writes a WRP message to the underlying response.  The format used is determined
 	// by the configuration of the underlying implementation.  This method is idempotent, and returns
 	// an error if called multiple times for the same instance.
-	WriteWRP(interface{}) (int, error)
+	WriteWRP(e *Entity) (int, error)
```
Comment (johnabass) [inline wrphttp/requestResponse.go:98]: This is the right idea.  But we also want a `WriteXXX(wrp.Format, []byte)` method. That allows for proxying and pass through, where a caller never deserialized the `[]byte` into a `wrp.Message`.

File: wrphttp/requestResponse.go [hunk 2]
```diff
@@ -105,16 +139,38 @@ type entityResponseWriter struct {
...
+func (erw *entityResponseWriter) WriteWRP(e *Entity) (int, error) {
+	if e == nil {
+		return 0, ErrUndefinedEntity
+	}
+
+	var output []byte
 
-	if err := encoder.Encode(v); err != nil {
-		return 0, err
+	if len(e.Bytes) > 0 && e.Format == erw.f {
```
Comment (kristinapathak) [inline wrphttp/requestResponse.go:144]: should we do a nil check of Bytes before trying to check its length?

Comment (joe94) [reply wrphttp/requestResponse.go:144]: I added tests for that case. Len of nil byte array == 0 so we should be good 😄

**Review by kristinapathak — APPROVED** [review]

File: wrphttp/requestResponse.go [hunk 3]
```diff
@@ -105,16 +139,38 @@ type entityResponseWriter struct {
 	f wrp.Format
 }
 
-func (erw *entityResponseWriter) WriteWRP(v interface{}) (int, error) {
-	var (
-		output  []byte
-		encoder = wrp.NewEncoderBytes(&output, erw.f)
-	)
+func (erw *entityResponseWriter) WriteWRP(e *Entity) (int, error) {
+	if e == nil {
```
Comment (johnabass) [inline wrphttp/requestResponse.go]: Generally, you don't need to do this.  It's ok to allow panics on nil.  That indicates a bug in the calling code, which you don't want to mask with a "normal" call return.

File: wrphttp/requestResponse.go [hunk 4]
```diff
@@ -105,16 +139,38 @@ type entityResponseWriter struct {
...
+		if err := encoder.Encode(e.Message); err != nil {
+			return 0, err
+		}
 	}
 
 	erw.ResponseWriter.Header().Set("Content-Type", erw.f.ContentType())
 	return erw.ResponseWriter.Write(output)
 }
+
+func (erw *entityResponseWriter) WriteWRPFromBytes(f wrp.Format, encodedWRP []byte) (int, error) {
```
Comment (johnabass) [inline wrphttp/requestResponse.go]: As a general naming convention, words like "from" are avoided in `golang` APIs.  This can be just `WriteWRPBytes`.

**Review by johnabass — CHANGES_REQUESTED** [review]

**Review by johnabass — APPROVED** [review]

**Pull request merged**